### PR TITLE
Added test for ICP example code

### PR DESCRIPTION
**Pull request by larshg**
As in an attempt to help with #3624, I have added a test for ICP - more or less copy of the [tutorial](https://github.com/PointCloudLibrary/pcl/blob/cc7fe363c6463a0abc617b1e17e94ab4bd4169ef/doc/tutorials/content/sources/iterative_closest_point/iterative_closest_point.cpp).

However, when I tried to check out pcl 1.8 I had (unrelated)linker errors and discontinued to investigate the error.

But maybe its of interest to have the ICP test?

### CODE REVIEW

**Comment by kunaltyagi** [comment]
@larshg That's a nice idea. Could you please add a few comments in the code to inform the reader where, why and what is being tested?

**Comment by kunaltyagi** [comment]
Maybe we need some generic method to confirm that the tutorial performs as expected, not just compile and link.

**Comment by larshg** [comment]
> Maybe we need some generic method to confirm that the tutorial performs as expected, not just compile and link

Do you mean generally or in this specific case?

Its validating that the found translation is within a threshold, which should indicate that the tutorial code performs as expected?

**Comment by kunaltyagi** [comment]
> Do you mean generally or in this specific case?

Generally. It might be possible that other tutorials too have their own minor issues.

File: test/registration/test_ICP.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  icp.align(Final);
+
+  // Check that we have sucessfully converged
+  ASSERT_EQ(icp.hasConverged(), true);
+
+  // Test that the fitness score is below acceptable threshold
+  EXPECT_LT(icp.getFitnessScore(), 1e-6);
+
+  // Ensure that the translation found is within acceptable threshold.
+  EXPECT_NEAR(icp.getFinalTransformation()(0, 3), 0.7, 2e-3);
```
Comment (taketwo) [inline test/registration/test_ICP.cpp]: I'd also check translations along other axes. Alternatively, check the norm of the entire translation.

File: test/registration/CMakeLists.txt [hunk 2]
```diff
@@ -19,6 +19,10 @@ PCL_ADD_TEST(registration_warp_api test_warps
 PCL_ADD_TEST(correspondence_estimation test_correspondence_estimation
              FILES test_correspondence_estimation.cpp
              LINK_WITH pcl_gtest pcl_registration pcl_features)
+            
+PCL_ADD_TEST(ICP test_ICP
+             FILES test_ICP.cpp
```
Comment (taketwo) [inline test/registration/CMakeLists.txt]: Please use lowercase in filenames and also test name.

**Comment by larshg** [comment]
Just realized there is a simliar test [here](https://github.com/PointCloudLibrary/pcl/blob/0e452dd9a19fc4a493772549c76b89615e20813a/test/registration/test_registration.cpp#L158-L192)

but is more or less disabled...

Should I move my test to registration?

And how is the supplied test files given to the test?

**Comment by kunaltyagi** [comment]
It seems like the 2 tests are almost the same. It might be better to enable the already present test and add any new checks there.

**Comment by taketwo** [comment]
> And how is the supplied test files given to the test?

CMake macro that creates a test associates argument filenames with it. So when the test executable is run by `ctest`/`make tests` it receives these very arguments.

https://github.com/PointCloudLibrary/pcl/blob/0e452dd9a19fc4a493772549c76b89615e20813a/test/registration/CMakeLists.txt#L27

**Comment by larshg** [comment]
Yeah, saw that - its just not propagated to a VS solution. Would it be troublesome to change it to just load relative paths instead? I guess its not often a unit test is changed with inputs?

**Comment by taketwo** [comment]
Awaiting #3636 merge.

**Comment by kunaltyagi** [comment]
Might be better to rebase to master instead of merging from a branch

**Comment by larshg** [comment]
Yea, I'm not sure why I got the other merged AND pushed...

But I'm rebasing and in the progress of enable the other tests. Though, I still need to find out why they got disabled in the first place.

**Comment by larshg** [comment]
I haven't tested the tutorial per se. Will do it asap.

**Comment by kunaltyagi** [comment]
#3636 has been merged. Did you test the tutorials?

**Comment by larshg** [comment]
Not yet, but I think I'll soon be able to do it!

**Comment by larshg** [comment]
Rebased and verified that the tutorial runs.

File: doc/tutorials/content/sources/iterative_closest_point/iterative_closest_point.cpp [hunk 3]
```diff
@@ -14,26 +14,32 @@ int
...
   }
+  
   std::cout << "Saved " << cloud_in->points.size () << " data points to input:"
       << std::endl;
-  for (std::size_t i = 0; i < cloud_in->points.size (); ++i) std::cout << "    " <<
-      cloud_in->points[i].x << " " << cloud_in->points[i].y << " " <<
-      cloud_in->points[i].z << std::endl;
+      
+  for (auto& point : cloud_in->points)
+    std::cout << "    " << point.x << " " << point.y << " " << point.z << std::endl;
```
Comment (SergioRAgostinho) [inline doc/tutorials/content/sources/iterative_closest_point/iterative_closest_point.cpp]:
```cpp
for (const auto& point : *cloud_in)
  std::cout << point << std::endl;
```
should be sufficient. The operator exists for `PointXYZ`.
https://github.com/PointCloudLibrary/pcl/blob/3f0114eb942e7a1688917587e9be1db51a57cb7e/common/src/point_types.cpp#L41-L46

File: doc/tutorials/content/sources/iterative_closest_point/iterative_closest_point.cpp [hunk 4]
```diff
@@ -14,26 +14,32 @@ int
...
+    point.x = point.x + 0.7f;
+
   std::cout << "Transformed " << cloud_in->points.size () << " data points:"
       << std::endl;
-  for (std::size_t i = 0; i < cloud_out->points.size (); ++i)
-    std::cout << "    " << cloud_out->points[i].x << " " <<
-      cloud_out->points[i].y << " " << cloud_out->points[i].z << std::endl;
+      
+  for (auto& point : cloud_out->points)
+    std::cout << "    " << point.x << " " << point.y << " " << point.z << std::endl;
```
Comment (SergioRAgostinho) [inline doc/tutorials/content/sources/iterative_closest_point/iterative_closest_point.cpp]: Same comment here.

File: test/registration/test_registration.cpp [hunk 5]
```diff
@@ -154,6 +154,53 @@ TEST (PCL, findFeatureCorrespondences)
...
+  pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_in(new pcl::PointCloud<pcl::PointXYZ>);
+  pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_out(new pcl::PointCloud<pcl::PointXYZ>);
+
+  // Fill in the CloudIn data
+  cloud_in->width = 5;
+  cloud_in->height = 1;
+  cloud_in->is_dense = false;
+  cloud_in->points.resize(cloud_in->width * cloud_in->height);
+
+  for (auto& point : cloud_in->points)
```
Comment (SergioRAgostinho) [inline test/registration/test_registration.cpp]:
```cpp
for (auto& point : *cloud_in)
```

File: test/registration/test_registration.cpp [hunk 6]
```diff
@@ -154,6 +154,53 @@ TEST (PCL, findFeatureCorrespondences)
...
+  {
+    point.x = 1024 * rand() / (RAND_MAX + 1.0f);
+    point.y = 1024 * rand() / (RAND_MAX + 1.0f);
+    point.z = 1024 * rand() / (RAND_MAX + 1.0f);
+  }
+
+  *cloud_out = *cloud_in;
+
+  for (auto& point : cloud_out->points)
```
Comment (SergioRAgostinho) [inline test/registration/test_registration.cpp]: Same comment.

File: test/registration/test_registration.cpp [hunk 7]
```diff
@@ -154,6 +154,52 @@ TEST (PCL, findFeatureCorrespondences)
...
+TEST(PCL, ICP_translated)
+{
+  pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_in(new pcl::PointCloud<pcl::PointXYZ>);
+  pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_out(new pcl::PointCloud<pcl::PointXYZ>);
+
+  // Fill in the CloudIn data
+  cloud_in->width = 5;
+  cloud_in->height = 1;
+  cloud_in->is_dense = false;
+  cloud_in->points.resize(cloud_in->width * cloud_in->height);
```
[the remark this answers is not in the document]
Comment (larshg) [reply test/registration/test_registration.cpp]: I don't really have any intentions, as it was more or less a copy of [the tutorial](https://github.com/PointCloudLibrary/pcl/blob/cc7fe363c6463a0abc617b1e17e94ab4bd4169ef/doc/tutorials/content/sources/iterative_closest_point/iterative_closest_point.cpp).

So we can probably change it as to your suggestion.

File: doc/tutorials/content/sources/iterative_closest_point/iterative_closest_point.cpp [hunk 8]
```diff
@@ -6,39 +6,40 @@
...
+  
   std::cout << "size:" << cloud_out->points.size() << std::endl;
-  for (std::size_t i = 0; i < cloud_in->points.size (); ++i)
-    cloud_out->points[i].x = cloud_in->points[i].x + 0.7f;
-  std::cout << "Transformed " << cloud_in->points.size () << " data points:"
-      << std::endl;
-  for (std::size_t i = 0; i < cloud_out->points.size (); ++i)
-    std::cout << "    " << cloud_out->points[i].x << " " <<
-      cloud_out->points[i].y << " " << cloud_out->points[i].z << std::endl;
+  for (auto& point : cloud_out->points)
```
Comment (SergioRAgostinho) [inline doc/tutorials/content/sources/iterative_closest_point/iterative_closest_point.cpp]:
```cpp
for (auto& point : *cloud_out)
```

**Pull request merged**